pkg/allocation/libs/converter.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2019 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package converters includes API concersions between GameServerAllocation API and the Allocation proto APIs
+package converters
```

Comment:
It looks like this was all done by hand. But the fields in the proto and the k8s api look very similar. Is there a reason that we don't auto-generate the service definition from the api spec (to reduce the amount of manual conversion code)?

Reply:
I looked into proteus. There are some issues generating proto from golang package due to old packages and mismatches, but it seems to be eventually doable. Now I am thinking how much value do we get by reusing Spec and Status. Here are the pros and cons for not auto generating part of API from GSA:

pros: 
 1. Separating external facing APIs and internal APIs for agones-allocator. The external API is allocation.proto and the internal API is GSA. Internal API may change, but the external API should not change after a stable release.
  2. The allocation.proto fields will follow best practices for REST and gRPC API design and GSA will follow the k8s API design. So for example in the case of status, I am planning to drop states (https://github.com/googleforgames/agones/issues/1040) for contention and unallocated and returning the expected gRPC errors. There are also field names that are different between the two e.g. preferredGameServerSelectors vs preferred.
  3. proto generation and keeping dependency in sync with proteus seems to be a hassle (I still haven't got it working)

Cons: 
1. There are more maintenance e.g. writing more code for conversion layer and making sure the API is in sync.

I am inclined to keep the APIs as is partly because of the benefits I explain which I feel outweights the disadvantages, and partly because I am biased at this stage :)
Let me know if you have a strong opinion on generating the proto.